searx/engines/pubmed.py

```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env python
+
+"""
+ PubMed (Scholar publications)
+
+ @website     https://www.ncbi.nlm.nih.gov/pubmed/
+ @provide-api yes (https://www.ncbi.nlm.nih.gov/home/develop/api/)
+
+ @using-api   yes
+ @results     XML
+ @stable      yes
+ @parse       url, title, publishedDate, content
+ More info on api: https://www.ncbi.nlm.nih.gov/books/NBK25501/
+"""
+
+from lxml import etree
+from datetime import datetime
+from searx.url_utils import urlencode
+try:
+    # For Python 3.0 and later
+    from urllib.request import urlopen
+except ImportError:
+    # Fall back to Python 2's urllib2
+    from urllib2 import urlopen
+
+
+categories = ['science']
+
+base_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi'\
+           + '?db=pubmed&{query}&retstart={offset}&retmax={hits}'
+
+# engine dependent config
+number_of_results = 10
+pubmed_url = 'https://www.ncbi.nlm.nih.gov/pubmed/'
+
+
+def request(query, params):
+    # basic search
+    offset = (params['pageno'] - 1) * number_of_results
+
+    string_args = dict(query=urlencode({'term': query}),
+                       offset=offset,
+                       hits=number_of_results)
+
+    params['url'] = base_url.format(**string_args)
+
+    return params
+
+
+def response(resp):
+    results = []
+
+    # First retrieve notice of each result
+    pubmed_retrieve_api_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?'\
+                              + 'db=pubmed&retmode=xml&id={pmids_string}'
+
+    # handle Python2 vs Python3 management of bytes and strings
+    try:
+        pmids_results = etree.XML(resp.text.encode('utf-8'))
+    except AttributeError:
+        pmids_results = etree.XML(resp.text)
+
+    pmids = pmids_results.xpath('//eSearchResult/IdList/Id')
+    pmids_string = ''
+
+    for item in pmids:
+        pmids_string += item.text + ','
+
+    retrieve_notice_args = dict(pmids_string=pmids_string)
+
+    retrieve_url_encoded = pubmed_retrieve_api_url.format(**retrieve_notice_args)
+
+    search_results_xml = urlopen(retrieve_url_encoded).read()
+    search_results = etree.XML(search_results_xml).xpath('//PubmedArticleSet/PubmedArticle/MedlineCitation')
+
+    for entry in search_results:
+        title = entry.xpath('.//Article/ArticleTitle')[0].text
+
+        # Optional: redirect directly to article page and if no DOI, redirect to PubMed
```

Comment:
Could you make the optional stuff configurable? So instead of commenting lines out, there could be a boolean which is `False` by default and can be configured from here or from `settings.yml`. This way the code can be kept cleaner, and those admins who don't like touching code can set it from config file. Also, if something changes in these lines and an admin who enabled this part has to deal with conflicts, too.

Reply:
I will make it configurable.

Are the parameters in the object

>  - name : pubmed
    parameter : True

accessible from the engine, _i.e._ callable without having to parse settings.yml and match the item with the good name.

Otherwise, I'll simply make it a parameter outside of the engine, but it will be less maintainable.

Reply:
I finally don't add a parameter in settings. Only display DOI when available. And if a plugin handles DOIs, then it can be made clickable.